sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -91,7 +91,11 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   private val tags: mutable.Map[TreeNodeTag[_], Any] = mutable.Map.empty
 
   protected def copyTagsFrom(other: BaseType): Unit = {
```

Comment:
thinking about it more, I think we missed one case of tree node tags. If the rule replaces one node with a new one, it makes sense to copy the tags. If the rule removes one or more nodes, it doesn't make sense to append the tags to an existing node.

However, it's too expensive to detect the removal action (you need to traverse the tree and see if the new node is a descendant of the original node). I think a compromise is
```
if (tags.isEmpty) {
  tags ++= other.tags
}
```

It's still possible that we copy tags to an existing node due to a node removal rule, but it's not a big issue if that node has no tags at all.

Reply:
looks good. @maryannxue any concerns with this approach ?

Reply:
Agree with @cloud-fan 's reasoning and proposed fix.

The only worry is that we're not 100% sure if we are breaking things somewhere else, but the tests don't say so yet. Can't think of an approach that would have less impact yet fix this problem perfectly. Let's go with this approach then.